Client/Frontend/Extensions/ClientPickerViewController.swift

```diff
@@ -172,91 +203,109 @@ class ClientPickerViewController: UITableViewController {
+            profile.remoteClientsAndTabs.getRemoteDevices() >>== { devices in
+                profile.getClients().uponQueue(.main) { result in
+                    guard let clients = result.successValue else { return }
+                    withExtendedLifetime(profile) {
+                        self.loadState = .Loaded
 
-                // If we are running from an app extension then make sure we shut down the profile as soon as we are
-                // done with it.
-                if self.profileNeedsShutdown {
-                    profile._shutdown()
-                }
+                        // If we are running from an app extension then make sure we shut down the profile as soon as we are
+                        // done with it.
+                        if self.profileNeedsShutdown {
+                            profile._shutdown()
+                        }
 
-                self.loadState = .Loaded
-                guard let c = result.successValue else {
-                    return
-                }
+                        self.loadState = .Loaded
 
-                self.updateClients(clients: c, endRefreshing: true)
+                        self.update(devices: devices, clients: clients, endRefreshing: true)
+                    }
+                }
             }
-        })
+        }
     }
 
-    fileprivate func updateClients(clients: [RemoteClient], endRefreshing: Bool) {
+    fileprivate func update(devices: [RemoteDevice], clients: [RemoteClient], endRefreshing: Bool) {
+        assert(Thread.isMainThread)
+        guard let profile = self.ensureOpenProfile() as? BrowserProfile, let account = profile.getAccount() else { return }
+
+        let fxaDeviceIds = devices.compactMap { $0.id }
+        let devices = devices.filter { fxaDeviceIds.contains($0.id ?? "") }
+        let newRemoteDevices = devices.filter { account.commandsClient.sendTab.isDeviceCompatible($0) }
+
+        func findClient(forDevice device: RemoteDevice) -> RemoteClient? {
+            return clients.find({ $0.fxaDeviceId == device.id })
+        }
+        let oldRemoteClients = devices.filter { !account.commandsClient.sendTab.isDeviceCompatible($0) }
+            .compactMap { findClient(forDevice: $0) }
+
+        let fullList = newRemoteDevices.sorted { $0.id ?? "" > $1.id ?? "" }.map { DeviceOrClient.device($0) }
+            + oldRemoteClients.sorted { $0.guid ?? "" > $1.guid ?? "" }.map { DeviceOrClient.client($0) }
+
         // Sort the lists, and compare guids and modified, to see if the list has changed and tableview needs reloading.
-        let isSame = clients.sorted(by: { $0.guid ?? "" > $1.guid ?? "" }).elementsEqual(self.clients.sorted(by: { $0.guid ?? "" > $1.guid ?? "" })) {
-            guard let g0 = $0.guid, let g1 = $1.guid else { return false }
-            return g0 == g1 && $0.modified == $1.modified
+        let isSame = fullList.elementsEqual(devicesAndClients) { new, old in
```

Comment:
I feel like it might be cleaner to just make our `DeviceOrClient` adhere to `Equatable`.